transaction/std/src/transaction_builder.rs

```diff
@@ -38,6 +38,22 @@ impl TxOutputsOrdering for DefaultTxOutputsOrdering {
     }
 }
 
+/// Transaction output context is produced by add_output method
+/// Used for receipt creation
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct TxOutContext {
+    /// TxOut that comes from a transaction builder
+    /// add_output/add_change_output
+    pub tx_out: TxOut,
+    /// confirmation that comes from a transaction builder
+    /// add_output/add_change_output
+    pub confirmation: TxOutConfirmationNumber,
+    /// Shared Secret that comes from a transaction builder
+    /// add_output/add_change_output
+    pub shared_secret: RistrettoPublic,
```

Comment:
This is a fairly sensitive piece of information, I'm not sure it should be coming out of transaction builder at all. The shared secret is derived via DH key exchange and can be derived on the other side/via ffi as long as the Moby app has the view private key.

At the very least it shouldn't be a public field.

Reply:
@iamalwaysuncomfortable this is required for Moby to do their thing, they need to encrypt additional stuff using this field.

I also see nothing wrong with it being a public field, this can just be an organizational type like `InputSecret` or `OutputSecret` with no particular invariants

Reply:
My issue with passing the shared secret directly is Moby isn't the only consumer of the tx builder and exposing it for all possible clients here seems to break security.

Wouldn't in be possible to add a method within the ffis that generates the shared secret from the tx_out_public key? And then passes it upwards through the boundary? There are currently other ffi methods calculating the shared secret.

Reply:
Also, I'll have to go digest that MCIP

Reply:
>  exposing it for all possible clients here seems to break security.

I guess I don't see why -- as long as they don't misuse it it's fine. The client already knows everything that this shared secret is concealing anyways.

I agree that this thing should be zeroized at the end but as long as that happens this isn't really a security issue. The clients compute the shared secret on their own for their own TxOut's all the time right now, it's part of the SDK interface.

Reply:
Yeah that's actually true, and James just pointed out it's in the same memory space. 

So: I'm sold. 

However, do we want to enforce that the TxBuilder add_output method now returns the shared secret as a tuple and the TxOutContext is constructed on the ffi's side via struct deconstruction? i.e. `let TxOutContext { tx_out, confirmation, .. } = tx_builder.add_output(...);`